### PR TITLE
Add "+" sign support for float exponent

### PR DESCRIPTION
**Pull request by mvorisek**
in scientific format `+` after `E` is normally present

also test if the parsed numbers are not converted thru native `float`/`int` type which can loose precision

### CODE REVIEW

**Comment by mvorisek** [comment]
PR is done

**Comment by mvorisek** [comment]
@ondrejmirtes I have rebased this PR from 1.9.x to 1.20.x and the first 3 PR seems to be missing in 1.20.x! Please check that:

image

**Comment by ondrejmirtes** [comment]
Because there's 1.21.x.

**Comment by mvorisek** [comment]
rebased and retargeted on 1.21.x

**Comment by mvorisek** [comment]
can this PR be merged?

**Comment by mvorisek** [comment]
This PR is done. Support for PHP CS Fixer for `+` has been already landed. "Test Slevomat Coding Standard" CI failures are unrelated.

File: src/Lexer/Lexer.php [hunk 1]
```diff
@@ -160,8 +160,8 @@ private function generateRegexp(): string
 			self::TOKEN_PHPDOC_TAG => '@(?:[a-z][a-z0-9-\\\\]+:)?[a-z][a-z0-9-\\\\]*+',
 			self::TOKEN_PHPDOC_EOL => '\\r?+\\n[\\x09\\x20]*+(?:\\*(?!/)\\x20?+)?',
 
-			self::TOKEN_FLOAT => '(?:-?[0-9]++(_[0-9]++)*\\.[0-9]*(_[0-9]++)*+(?:e-?[0-9]++(_[0-9]++)*)?)|(?:-?[0-9]*+(_[0-9]++)*\\.[0-9]++(_[0-9]++)*(?:e-?[0-9]++(_[0-9]++)*)?)|(?:-?[0-9]++(_[0-9]++)*e-?[0-9]++(_[0-9]++)*)',
-			self::TOKEN_INTEGER => '-?(?:(?:0b[0-1]++(_[0-1]++)*)|(?:0o[0-7]++(_[0-7]++)*)|(?:0x[0-9a-f]++(_[0-9a-f]++)*)|(?:[0-9]++(_[0-9]++)*))',
+			self::TOKEN_FLOAT => '(?:[+-]?[0-9]++(_[0-9]++)*+\\.[0-9]*+(_[0-9]++)*+(?:e[+-]?[0-9]++(_[0-9]++)*+)?)|(?:[+-]?[0-9]*+(_[0-9]++)*+\\.[0-9]++(_[0-9]++)*+(?:e[+-]?[0-9]++(_[0-9]++)*+)?)|(?:[+-]?[0-9]++(_[0-9]++)*+e[+-]?[0-9]++(_[0-9]++)*+)',
```
Comment (ondrejmirtes) [inline src/Lexer/Lexer.php]: Some of these changes seem random to me and I don't know what to think about them. For example I don't know why there needs to be `+` after `*` - the last change on this line.

Your description says you want to add support for `+` after `e`. If so then please just add and test that.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Lexer/Lexer.php (hunk 1 above)
Comment (mvorisek) [reply src/Lexer/Lexer.php]: > `+` after `*`

I added it to help PCRE performance (in theory) as `++`/`*+` was already present. Is that ok?

> Your description says you want to add support for `+` after `e`. If so then please just add and test that.

This PR adds `+` sign support before any number (float/int) and after `E` as well. Is any test missing?

Comment (ondrejmirtes) [reply src/Lexer/Lexer.php]: Please just add support for `e+` and delete all the other changes.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Lexer/Lexer.php (hunk 1 above)
Comment (mvorisek) [reply src/Lexer/Lexer.php]: Fixed.

File: src/Lexer/Lexer.php [hunk 2]
```diff
@@ -160,7 +160,7 @@ private function generateRegexp(): string
 			self::TOKEN_PHPDOC_TAG => '@(?:[a-z][a-z0-9-\\\\]+:)?[a-z][a-z0-9-\\\\]*+',
 			self::TOKEN_PHPDOC_EOL => '\\r?+\\n[\\x09\\x20]*+(?:\\*(?!/)\\x20?+)?',
 
-			self::TOKEN_FLOAT => '(?:-?[0-9]++(_[0-9]++)*\\.[0-9]*(_[0-9]++)*+(?:e-?[0-9]++(_[0-9]++)*)?)|(?:-?[0-9]*+(_[0-9]++)*\\.[0-9]++(_[0-9]++)*(?:e-?[0-9]++(_[0-9]++)*)?)|(?:-?[0-9]++(_[0-9]++)*e-?[0-9]++(_[0-9]++)*)',
+			self::TOKEN_FLOAT => '(?:-?[0-9]++(_[0-9]++)*\\.[0-9]*+(_[0-9]++)*+(?:e[+-]?[0-9]++(_[0-9]++)*)?)|(?:-?[0-9]*+(_[0-9]++)*\\.[0-9]++(_[0-9]++)*(?:e[+-]?[0-9]++(_[0-9]++)*)?)|(?:-?[0-9]++(_[0-9]++)*e[+-]?[0-9]++(_[0-9]++)*)',
```
Comment (mvorisek) [inline src/Lexer/Lexer.php:163]: one `+` was missing here for consistency, so I added it

File: tests/PHPStan/Parser/ConstExprParserTest.php [hunk 3]
```diff
@@ -152,6 +152,11 @@ public function provideIntegerNodeParseData(): Iterator
 			new ConstExprIntegerNode('123'),
 		];
 
+		yield [
+			'-123',
+			new ConstExprIntegerNode('-123'),
```
Comment (mvorisek) [inline tests/PHPStan/Parser/ConstExprParserTest.php:157]: negative integer test was missing

File: tests/PHPStan/Parser/ConstExprParserTest.php [hunk 4]
```diff
@@ -227,8 +237,8 @@ public function provideFloatNodeParseData(): Iterator
 		];
 
 		yield [
-			'-123',
-			new ConstExprIntegerNode('-123'),
+			'-123.',
+			new ConstExprFloatNode('-123.'),
```
Comment (mvorisek) [inline tests/PHPStan/Parser/ConstExprParserTest.php:241]: this was definitely typo, float should be tested here

File: src/Lexer/Lexer.php (hunk 2 above)
Comment (ondrejmirtes) [inline src/Lexer/Lexer.php]: I feel like there are more `e-` changes than added tests. Can you please make sure that you have a failing test first for each `[+-]` added? Thanks.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Lexer/Lexer.php (hunk 2 above)
Comment (mvorisek) [reply src/Lexer/Lexer.php]: done

**Comment by ondrejmirtes** [comment]
Thank you. Feel free to submit another PR that allows `+` at the beginning of numbers. I feel like these are two separate problems so they should be handled separately. It felt out of place in this PR but I just noticed that `doctrine/annotations` also support that (and I'm currently adding support here to be also able to parse the doctrine/annotations syntax) so this will come in handy. Thank you once more.

**Pull request merged**